flask_rebar/utils.py

```diff
@@ -0,0 +1,124 @@
+"""
+    General-Purpose Utilities
+    ~~~~~~~~~~~~~~~~~~~~~~~~~
+
+    Utilities that are not specific to request-handling (you'll find those in request_utils.py).
+
+    :copyright: Copyright 2019 Autodesk, Inc., see AUTHORS.
+    :license: MIT, see LICENSE for details.
+"""
+
+import functools
+import warnings
+
+from werkzeug.local import LocalProxy as module_property  # noqa
+
+
+# ref http://jtushman.github.io/blog/2014/05/02/module-properties/ for background on
+# use of werkzeug LocalProxy to simulate "module properties"
+# end result: singleton config can be accessed by, e.g.,
+#    from flask_rebar.utils import deprecation_config
+#    deprecation_config.warning_type = YourFavoriteWarning
+
+
+class DeprecationConfig:
+    """
+    Singleton class to allow one-time set of deprecation config controls
+    """
+
+    __instance = None
+
+    @staticmethod
+    def getInstance():
+        """ Static access method. """
+        if DeprecationConfig.__instance is None:
+            DeprecationConfig()
+        return DeprecationConfig.__instance
+
+    def __init__(self):
+        """ Virtually private constructor. """
+        if DeprecationConfig.__instance is not None:
+            raise Exception("This class is a singleton!")
+        else:
+            DeprecationConfig.__instance = self
+            self.warning_type = FutureWarning
+
+
+@module_property
+def deprecation_config():
+    return DeprecationConfig.getInstance()
+
+
+def deprecated(new_func=None, eol_version=None):
+    """
+    :param Union[str, (str, str)] new_func: Name (or name and end-of-life version) of replacement
+    :param str eol_version: Version in which this function may no longer work
+    :return:
+    """
+
+    def decorator(f):
+        @functools.wraps(f)
+        def wrapper(*args, **kwargs):
+            new = None
+            eol = eol_version
+            if (
+                type(new_func) is tuple
+            ):  # in case somebody inferred this from the way we deprecate params..
```

Comment:
If you are going to do this you probably should also check that `eol_version is None`.